myapp/app/models/user.rb

```diff
@@ -3,8 +3,22 @@
 # some comments for user model
 class User < ApplicationRecord
   has_secure_password
+  before_destroy :at_least_one_admin_remain
+
+  enum role: {
+    admin: 0,
+    normal: 1,
+  }
+
   has_many :tasks, dependent: :destroy
   validates :name, presence: true, length: { maximum: 255 }
   validates :password, presence: true, length: { maximum: 255 }, on: :create
   validates :description, length: { maximum: 1000 }
+  validates :role, presence: true
+
+  private
+
+  def at_least_one_admin_remain
+    throw(:abort) if User.admin.length == 1 && self.admin?
```

Comment:
If you write it like this, the SQL query won't be heavy because the search will be done with `SELECT COUNT 1`.

```suggestion
    throw(:abort) if User.exists?(role: :admin) == 1 && self.admin?
```

Reply:
Thanks for the kind comment about the sql query optimization!
But in this case, I think only `exists?` is not enough because the deletion should be approved if there are more than 1 admin users. If and only if there is 1 admin user, and that user is going to be deleted, then the delete operation should be rejected. If there are 2 or more admin users, it can be deleted and we will still get admin user(s) remained.